Single-machine multi-node mixed cluster CI prerequisites

deepfire opened the pull request:
This supplies the necessary changes for a mixed-cluster integration test, as per https://github.com/input-output-hk/cardano-node/issues/255 :

1. `mainnet_ci_full` genesis & configuration, starting in OBFT node
2. fix for an OBFT EBB rollback issue, which was trying to erase EBB even if the chain was started in OBFT mode, leading to https://github.com/input-output-hk/cardano-node/issues/255#issuecomment-543247537
3. change in `network-transport-tcp` to be more lenient regarding remote address claims: https://github.com/deepfire/network-transport-tcp/commit/44f84a887cf954dd924f28b7e8da107228e57c13.  This is necessary to avoid problems when starting multiple nodes on the same machine.
4. small improvements in genesis generation

Additionally, this resets the protocol version for the `shelley_staging_short_full` configuration to 0 -- a prerequisite for its respin.

*NOTE*: perhaps this PR should be split.  But then, this repository sees very little activity, so perhaps the separation wouldn't have much benefit.  I don't have a strong opinion myself.

intricate commented on lib/src/Pos/Worker/Block.hs:
```diff
@@ -211,7 +211,7 @@ dropObftEbb genesisConfig txpConfig = do
     case tipHeader of
         BlockHeaderMain _      -> pure ()
-        BlockHeaderGenesis _   -> do
+        BlockHeaderGenesis _   -> unless (tipHeader ^. epochIndexL == 0) $ do
```
We could potentially doc why this `unless` is necessary. I guess a comment similar to your commit message might be sufficient. Maybe something like:

```
-- If we're starting up a chain in OBFT mode, we need to ensure that we don't 
-- rollback the actual genesis block out of existence (i.e. the EBB at epoch 0)
```

What do you think?

deepfire replied:
Good point, done @intricate !

intricate commented on stack.yaml:
```diff
@@ -94,8 +94,8 @@ packages:
-    git: https://github.com/avieth/network-transport-tcp
-    commit: 2634e5e32178bb0456d800d133f8664321daa2ef # csl-0.6.0
+    git: https://github.com/deepfire/network-transport-tcp
+    commit: 44f84a887cf954dd924f28b7e8da107228e57c13 # csl-0.6.0
```
> change in network-transport-tcp to be more lenient regarding remote address claims

In `44f84a887cf954dd924f28b7e8da107228e57c13`, is there good reason for dropping the `theirHost == numericHost || theirHost == resolvedHost` check?

deepfire replied:
It was giving problems when starting a bunch of nodes on the same machine -- and I went to extra lengths to try avoiding this:

1. setting up a VDE2 [1] virtual LAN, with one TAP interface per node (and hence a dedicated non-loopback IP)
2. Implementing source policy routing [2]

This got me quite far regarding the address inconsistencies that that error check was giving me, but alas, I still couldn't get past.

There is a suspicion that it probably wasn't critical anyway -- but perhaps we want to double-check with @avieth.

--
1. https://github.com/virtualsquare/vde-2
2.  https://www.tldp.org/HOWTO/Adv-Routing-HOWTO/lartc.rpdb.simple.html

deepfire replied:
Dropped the `network-transport-tcp` changes & replaced them with a CLI flag driving the parameter that @avieth pointed out.

avieth commented:
> change in network-transport-tcp to be more lenient regarding remote address claims: deepfire/network-transport-tcp@44f84a8. This is necessary to avoid problems when starting multiple nodes on the same machine.

What exactly is the problem? That patch to nt-tcp opens up a denial of service attack. The check was put there for good reason.

avieth commented:
If you really _must_ disable the hostname check, you can set it in `TCPParameters` http://hackage.haskell.org/package/network-transport-tcp-0.6.0/docs/Network-Transport-TCP.html#t:TCPParameters

But beware! Your node will be vulnerable.

deepfire commented:
@avieth, thank you!  I've missed the check at https://github.com/avieth/network-transport-tcp/blob/csl-2.0.0/src/Network/Transport/TCP.hs#L1044

deepfire commented:
@avieth, the example failure:
```
Sep 09 01:46:48 a-node xadsbs7wlnm94wv14v9w8c0ba1dpjsyw-unit-script-cardano-node-legacy_-start[22055]: [cardano-sl.diffusion.outboundqueue.c-a-2:Warning:ThreadId 150] [2001-09-09 01:46:48.23 UTC] sending MsgAnnounceBlockHeader OriginSender to NodeId 10.1.0.3:3000:0 failed with TransportError ConnectFailed "setupRemoteEndPoint: Host mismatch. Claimed: 10.1.0.2; Numeric: 10.1.0.3; Resolved: c-b-1.cardano" :: SomeException
Sep 09 01:46:48 a-node xadsbs7wlnm94wv14v9w8c0ba1dpjsyw-unit-script-cardano-node-legacy_-start[22055]: [cardano-sl.diffusion.outboundqueue.c-a-2:Warning:ThreadId 149] [2001-09-09 01:46:48.23 UTC] sending MsgAnnounceBlockHeader OriginSender to NodeId 10.1.0.1:3000:0 failed with TransportError ConnectFailed "setupRemoteEndPoint: Host mismatch. Claimed: 10.1.0.2; Numeric: 10.1.0.1; Resolved: c-a-1.cardano" :: SomeException
Sep 09 01:46:48 a-node xadsbs7wlnm94wv14v9w8c0ba1dpjsyw-unit-script-cardano-node-legacy_-start[22055]: [cardano-sl.diffusion.outboundqueue.c-a-2:Warning:ThreadId 155] [2001-09-09 01:46:48.23 UTC] sending MsgAnnounceBlockHeader OriginSender to NodeId 10.1.0.7:3000:0 failed with TransportError ConnectFailed "setupRemoteEndPoint: Host mismatch. Claimed: 10.1.0.2; Numeric: 10.1.0.7; Resolved: c-d-1.cardano" :: SomeException
```

```
[root@a-node:~]# netstat -pltn | grep cardano
tcp        0      0 10.1.0.2:3000           0.0.0.0:*               LISTEN      22055/cardano-node- 
tcp        0      0 10.1.0.5:3000           0.0.0.0:*               LISTEN      22078/cardano-node- 
tcp        0      0 10.1.0.6:3000           0.0.0.0:*               LISTEN      22075/cardano-node- 
tcp        0      0 10.1.0.7:3000           0.0.0.0:*               LISTEN      22068/cardano-node- 
tcp        0      0 10.1.0.3:3000           0.0.0.0:*               LISTEN      22092/cardano-node- 
tcp        0      0 10.1.0.4:3000           0.0.0.0:*               LISTEN      22083/cardano-node- 
tcp        0      0 10.1.0.1:3000           0.0.0.0:*               LISTEN      22053/cardano-node- 
```

```
[root@a-node:~]# host 10.1.0.1
1.0.1.10.in-addr.arpa domain name pointer c-a-1.cardano.

[root@a-node:~]# host 10.1.0.2
2.0.1.10.in-addr.arpa domain name pointer c-a-2.cardano.

[root@a-node:~]# host 10.1.0.3
3.0.1.10.in-addr.arpa domain name pointer c-b-1.cardano.

[root@a-node:~]# host 10.1.0.4
4.0.1.10.in-addr.arpa domain name pointer c-b-2.cardano.
```

deepfire commented:
@avieth, 

1. dropped the `network-transport-tcp` change
2. added a CLI option to drive the host peer address consistency checking -- accompanied with sufficiently scary message during startup & comments

deepfire commented:
bors r+

deepfire commented:
bors r-

mhuesch commented on lib/src/Pos/Worker/Block.hs:
```diff
@@ -211,7 +211,10 @@ dropObftEbb genesisConfig txpConfig = do
+        -- If we're starting up a chain in OBFT mode, we need to ensure
+        -- that we don't rollback the actual genesis block out of existence
+        -- (i.e. the EBB at epoch 0):
+        BlockHeaderGenesis _   -> unless (tipHeader ^. epochIndexL == 0) $ do
```
Yes, this looks completely reasonable. My memory is fuzzy, but I thought we had tested scenarios where the cluster started in OBFT mode, but perhaps I am misremembering - it seems like we would've hit the same snag that Serge encountered here.

intricate replied:
@mhuesch: Yeah, we definitely had tested those scenarios before but I think after implementing your rollback solution we eventually ended up accepting a trade-off here such that we'd not be able to start up a chain directly in OBFT.

deepfire commented:
bors r+
